### ISSUE TITLE
Fix TypeScript error to export `Plugin` and `RuleContext`

### ISSUE THREAD

**Issue by henryruhs** [stylelint/stylelint#6662]
### What steps are needed to reproduce the bug?

As I wrote my plugin 100% in TypeScript (which is properly not intended by the Stylelint team) I used couple of internal types that are no longer exported such as `Plugin` and `RuleContext`. For this I had to create a `polyfile.type.ts` to work around it.

Other than that, the `lint`, `utils`, `createPlugin` and properly all of the PublicAPI members need to be imported this way.

```
import stylelint from 'stylelint';

const { lint, utils, createPlugin } : typeof stylelint = stylelint;
```

I could not even use `PublicApi` as typedef as this is not exported aswell - therfore the `typeof styleint` hack.

### What Stylelint configuration is needed to reproduce the bug?

Checkout the [master](https://github.com/isnotdefinedcom/stylelint-plugin/tree/master) of my Stylelint 14 plugin and upgrade the devDependencies to Stylelint 15. Once npm install is done, you can run npm test to see a bunch of errors.

### How did you run Stylelint?

npm test

### Which version of Stylelint are you using?

15.1.0

### What did you expect to happen?

...

### What actually happened?

You can see all my workarounds in the [develop](https://github.com/isnotdefinedcom/stylelint-plugin/tree/develop) branch of my plugin or study the [changes](https://github.com/isnotdefinedcom/stylelint-plugin/compare/master...develop) via aswell.

### Proposal to fix the bug

There is the need to export members that are used by plugin authors.

**Comment by ybiquitous** [comment]
@henryruhs Thanks for the report with the details. Your proposal sounds good to me. 👍🏼 

When seeing [your type polyfill](https://github.com/isnotdefinedcom/stylelint-plugin/blob/ba7f9c209604fff6507b0140f9f1d2c3d86e60fb/src/polyfill.type.ts), do you want to export the `Plugin` and `RuleContext` types, right? If so, we can export the followings:

https://github.com/stylelint/stylelint/blob/063d25d91227d2cd38980677fb98201761f1e75b/types/stylelint/index.d.ts#L12

https://github.com/stylelint/stylelint/blob/063d25d91227d2cd38980677fb98201761f1e75b/types/stylelint/index.d.ts#L183

It would be preferred to add JSDoc at the same time as exporting.

I've labeled the issue as **ready to implement**. Please consider [contributing](https://stylelint.io/contributing) if you have time.

**Comment by henryruhs** [comment]
I just send a pull request with the needed changes - something is off with the changeset... sorry for not having the time to get into that.

In general I would suggest to refactor your types.

1. Types marked as internal or even named `InternalApi` are exported
2. On the other hand the `PublichApi` is not exported and contains a internal property marked with underscore

I recommend the following:

1. to rename the current index.d.ts to types.d.ts, export all types and remove declare module
2. create a new index.d.ts and export everything that is public within declare module
3. use the TypeScript utilitites for types: https://www.typescriptlang.org/docs/handbook/utility-types.html

**Comment by ybiquitous** [comment]
@henryruhs Your suggestion sounds good to me. If so, plugin authors will only be able to *real* public APIs or types, right?

Also, will our internal type checking (`npm run lint:types`) continue to work?

**Comment by henryruhs** [comment]
Pretty sure it will continue to work. The idea is that you guys don't have to limit yourself while developing internals and can pick the public types later on. Usually a manual declare of module is not needed if you use something like rollup.js that does the job for you.

**Comment by ybiquitous** [comment]
Great. Will you create another PR for the refactoring after merging #6664 if you have time, @henryruhs?

**Comment by henryruhs** [comment]
I could spend some time on refactoring your types, just needs some alignment. Chat or GitHub issue?

**Comment by jeddy3** [comment]
GitHub issue, please. We use issues as our one means as comms as we're in different time zones and prefer async.
